### PR TITLE
:sparkles: Add dangerous workflow check with untrusted code checkout pattern

### PR DESCRIPTION
**Pull request by asraa**
Signed-off-by: Asra Ali <asraa@google.com>

* **Please check if the PR fulfills these requirements**
- [ x] Tests for the changes have been added (for bug fixes / features)
- [ x] PR title follows the guidelines defined in  https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Adds a new check `Dangerous-Workflow` that checks if a github workflow uses dangerous coding patterns. 
The first pattern implemented is a check for an untrusted code checkout. 

This is described in https://github.com/ossf/scorecard/issues/426

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Other information**:

### CODE REVIEW

**Comment by azeemshaikh38** [comment]
Given recent discussion about Scorecard data change affecting clients, let's disable this check from running by default in CLI and cron job for now. If clients specifically want to invoke this check, that's ok.

**Comment by laurentsimon** [comment]
> Given recent discussion about Scorecard data change affecting clients, let's disable this check from running by default in CLI and cron job for now. If clients specifically want to invoke this check, that's ok.

@asraa use an env variable to guard the access to this check. You can update the code in `cmd/root.go`, function `getEnabledChecks()`

File: checks/dangerous_workflow.go
Comment (azeemshaikh38) [reply checks/dangerous_workflow.go:72]: Let's not block this PR on the parser discussion. We can discuss the parser issue in our upcoming meet and update code accordingly in a separate PR.

**Comment by azeemshaikh38** [comment]
> > Given recent discussion about Scorecard data change affecting clients, let's disable this check from running by default in CLI and cron job for now. If clients specifically want to invoke this check, that's ok.
> 
> @asraa use an env variable to guard the access to this check. You can update the code in `cmd/root.go`, function `getEnabledChecks()`

Just `delete(enabledchecks, checks.CheckDangerousWorkflow)` [here](https://github.com/ossf/scorecard/blob/83649a799ec4a7953a3180297311c4427edd0474/cmd/root.go#L120) should be enough  :) 
Same for cron job [here](https://github.com/ossf/scorecard/blob/83649a799ec4a7953a3180297311c4427edd0474/cron/worker/main.go#L196).

**Comment by laurentsimon** [comment]
> > > Given recent discussion about Scorecard data change affecting clients, let's disable this check from running by default in CLI and cron job for now. If clients specifically want to invoke this check, that's ok.
> > 
> > 
> > @asraa use an env variable to guard the access to this check. You can update the code in `cmd/root.go`, function `getEnabledChecks()`
> 
> Just `delete(enabledchecks, checks.CheckDangerousWorkflow)` [here](https://github.com/ossf/scorecard/blob/83649a799ec4a7953a3180297311c4427edd0474/cmd/root.go#L120) should be enough :) 

I want to be able to use and test the local repo interface with all checks (without providing a long list of `--checks`), so I'd prefer an env variable. 
 
Same for cron job [here](https://github.com/ossf/scorecard/blob/83649a799ec4a7953a3180297311c4427edd0474/cron/worker/main.go#L196).

**Comment by azeemshaikh38** [comment]
> I want to be able to use and test the local repo interface with all checks (without providing a long list of `--checks`), so I'd prefer an env variable.
> 
> Same for cron job [here](https://github.com/ossf/scorecard/blob/83649a799ec4a7953a3180297311c4427edd0474/cron/worker/main.go#L196).

If its only needed for testing, you can simply remove the line locally? ;)

File: checks/dangerous_workflow.go
Comment (chrismcgehee) [reply checks/dangerous_workflow.go:72]: Agreed. Let's merge this PR and later consolidate our parsing with the library Naveen's suggested.

**Comment by asraa** [comment]
I went with the ENV var solution, but only because I imagine future checks might just want to add to the list? And it would centralize where the code could be removed before the next cut but let me know if I should change that

File: checks/dangerous_workflow.go
Comment (naveensrinivasan) [reply checks/dangerous_workflow.go:72]: I concur, Let's merge this PR. Thanks

**Comment by laurentsimon** [comment]
@asraa heads up. https://github.com/ossf/scorecard/pull/1195 got merged. You're going to have to update the checks.yaml file and add a `repos` field with a list of interfaces that are supported by your check. See example [checks.yaml#L78](https://github.com/ossf/scorecard/blob/main/docs/checks/internal/checks.yaml#L78).

The doc generation will also validate that your check does not use APIs that are inconsistent with the `repos` field. If you find a problem with the validation, let me know. (Note: validation happens automatically as you re-generate the .md)

File: checks/dangerous_workflow.go
Comment (chrismcgehee) [inline checks/dangerous_workflow.go:54]: Hey Asra, some of the repos we check have random shell or python files in the .github/workflows directory. It would be good to check if this is a yaml file like in [pinned_dependencies.go](https://github.com/ossf/scorecard/blob/6562cc1f4488c7a018b5c6b4e031f990058d95a2/checks/pinned_dependencies.go#L525-L527). I'm creating a PR to fix this same issue in Token-Permissions.

File: docs/checks/internal/checks.yaml
Comment (chrismcgehee) [inline docs/checks/internal/checks.yaml:654]: A lot of the information on this advisory for the `check-spelling` Github Action won't be relevant to people needing to remediate their workflow. Can we link to https://securitylab.github.com/research/github-actions-preventing-pwn-requests/ instead?

File: checks/dangerous_workflow.go
Comment (chrismcgehee) [inline checks/dangerous_workflow.go:194]: I don't think this is accurate for PR workflows. From https://github.com/actions/checkout#usage:
```
# The branch, tag or SHA to checkout. When checking out the repository that
# triggered a workflow, this defaults to the reference or SHA for that event.
# Otherwise, uses the default branch.
ref: ''
```
Sounds to be that when a PR triggers the workflow, then the latest commit in that PR will be checked out.

Comment (laurentsimon) [reply checks/dangerous_workflow.go:194]: How about tweaking the comment to "In the context of a `pull_request_target` event, the absence of `ref` checks out the target branch of the pull request" or something to this effect...?

Comment (asraa) [reply checks/dangerous_workflow.go:194]: I'm trying to read up on this, but I believe actually it checks out the base of the PR which is trusted
(See this thread https://github.community/t/can-workflow-changes-be-used-with-pull-request-target/178626

and GITHUB_REF in https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#pull_request_target)

File: docs/checks/internal/checks.yaml
Comment (asraa) [reply docs/checks/internal/checks.yaml:654]: that makes sense, done!

File: checks/dangerous_workflow.go
Comment (asraa) [reply checks/dangerous_workflow.go:54]: done!

**Comment by asraa** [comment]
thanks for the review @chrismcgehee!

File: checks/dangerous_workflow.go
Comment (laurentsimon) [reply checks/dangerous_workflow.go]: Good catch. Im curious then: what if the base PR is cloned from a non-Branch-Protected branch on the original repo? Is it considered trusted?
Just curious... if there were testing workflows in that branch that auto-merge, etc it could be dangerous too.. probably not a very common threat anyway?

Comment (asraa) [reply checks/dangerous_workflow.go]: Right -- I think that if the base branch wasn't branch protected that would be a problem as well. But I assume that you should be using BP check in conjunction

Comment (laurentsimon) [reply checks/dangerous_workflow.go]: I mean, main branch is protected, but branch used by the PR is not. Say:
- repo has branch `main` and `test`. `main` is protected, but `test` is not.
- attacker clone the repo and update `test` branch locally.
- attacker creates PR to merge its local cloned `test` into our `main` branch.

Is this situation possible? If so, is there a way to detect it?

Comment (asraa) [reply checks/dangerous_workflow.go]: > attacker creates PR to merge its local cloned test into our main branch.

But then the target branch would be the main branch -- so the code being checked out in pull_request_target, GITHUB_REF, would be the main, not the untrusted code.

Comment (laurentsimon) [reply checks/dangerous_workflow.go]: ah, so I mis-understood the link. My original understanding was correct then. Sorry for the noise, ignore this comment :-)

**Comment by asraa** [comment]
ping? anything else to do here?

**Comment by laurentsimon** [comment]
> ping? anything else to do here?

license check is failing. apart from that, I think we're good to merge. If anyone disagrees, please reply.

**Comment by asraa** [comment]
oops, something funky with the merge -- one second will update

**Comment by olivekl** [comment]
> > ping? anything else to do here?
> 
> license check is failing. apart from that, I think we're good to merge. If anyone disagrees, please reply.

I'd just had a couple comments above on the documentation description. We could also address them after the merge, though.

**Comment by laurentsimon** [comment]
> > > ping? anything else to do here?
> > 
> > 
> > license check is failing. apart from that, I think we're good to merge. If anyone disagrees, please reply.
> 
> I'd just had a couple comments above on the documentation description. We could also address them after the merge, though.

I think there's time to add your suggestions in this PR, since the pre-submit tests are still failing.

**Comment by asraa** [comment]
Yeah! Happy to address more comments -- merge issue fixed

**Comment by laurentsimon** [comment]
did you update the cron job https://github.com/ossf/scorecard/blob/83649a799ec4a7953a3180297311c4427edd0474/cron/worker/main.go#L196?

**Comment by asraa** [comment]
> did you update the cron job

Oops -- good catch, I had assumed that it would call the root command, but it doesn't. Updated

**Review by olivekl — COMMENTED** [review]
Just a few comments!

File: docs/checks.md
Comment (asraa) [reply docs/checks.md]: Yay thanks!!

> "This check determines whether the project has dangerous code patterns. ADD: Some examples of these patterns are [X] and [Y]."

added

> Is there a second part to the check that I'm not seeing?

Yep, I've only added one so far but the plan is to add a series of checks (https://github.com/ossf/scorecard/issues/426#issuecomment-856939499). I tried to rephrase this with `The first code pattern checked`. So I think it makes sense to keep this more general.

>  Or is this test only applicable to GitHub (in which case, let's add that)?

The check will have a lot of GitHub Actions dependent style and relies on it's particular default behavior, so I added this clarification ` Determines if the project's GitHub Action workflows avoid dangerous patterns.`

File: docs/checks.md
Comment (laurentsimon) [inline docs/checks.md:255]: this check is only supported on GitHub. @olivekl do we need to use `merge request` instead of `pull request` in this sentence?

**Comment by laurentsimon** [comment]
I fixed the pre-submit error 
```
--- FAIL: TestSARIFOutput (0.00s)
    --- FAIL: TestSARIFOutput/check-8 (0.00s)
        sarif_test.go:800: check-8: invalid result: 1
```
in https://github.com/ossf/scorecard/pull/1233
Once you push the error should go away

File: docs/checks.md
Comment (olivekl) [reply docs/checks.md:255]: Since it's only on GitHub I think `pull request` is fine.

**Comment by asraa** [comment]
ping? anything else to do here? 
(assuming I'll update this to use the github workflow parser next)

**Review by olivekl — APPROVED** [review]

**Comment by laurentsimon** [comment]
> ping? anything else to do here? (assuming I'll update this to use the github workflow parser next)

nop, go ahead update with workflow parser. Unless it's a trivial change, do it in a follow-up PR.

**Comment by azeemshaikh38** [comment]
> > ping? anything else to do here? (assuming I'll update this to use the github workflow parser next)
> 
> nop, go ahead update with workflow parser. Unless it's a trivial change, do it in a follow-up PR. Let me know which option you prefer.

+1

**Comment by dota17** [comment]
> 
It is recommended that the Readme  of  the project be updated synchronously. Currently, there are 17 check items, whereas the Readme  has only 16 check items.
https://github.com/ossf/scorecard/blob/main/README.md

**Comment by naveensrinivasan** [comment]
Thanks, we would appreciate if you can do PR to fix the issue.

**Pull request merged**
